File: LeetCode_Algorithm/47(Permutations II)/47.py
from typing import List

'''
LeetCode 47 Permutation 2
중복을 제거한 perm 만들기
sol1 T : 140ms 41% S : 14.3MB 50%
숫자 카운트를 먼저 한 뒤 그 카운트에서 하나씩 빼가면서 
그 숫자가 1이상이라면 카운트를 줄이고 뎁스를 내려감
'''
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        # sol1 
        ans = []
        from collections import Counter
        counter = Counter(nums)
        def solve(path):
            if len(path) == len(nums):
                ans.append(path)
                return 
            for k in counter:
                if counter[k] > 0:
                    counter[k] -= 1
                    solve(path + [k])
                    counter[k] += 1
        path = []
        solve(path)
        return ans
        

            

print(Solution().permuteUnique([1,1,3]))

File: LeetCode_Algorithm/47(Permutations II)/47.js
var solve = function(path, counter, length) {
  if (path.length == length) {
    ans.push(path); return ;
  }
  counter.forEach((_,k) => {
    if (counter.get(k) > 0) {
      counter.set(k, counter.get(k)-1)
      path.push(k)
      solve([...path], counter, length)
      path.pop()
      counter.set(k, counter.get(k)+1)
    }
  })
}
var permuteUnique = function(nums) {
  // sol1 T : 79ms 96% S : 45.6MB 56.4%
  let hashMap = new Map()
  for (let num of nums) {
    if (hashMap.has(num)) {
      hashMap.set(num, hashMap.get(num)+1)
    } else {
      hashMap.set(num, 1)
    }
  }
  let path = []; ans = []
  solve(path, hashMap, nums.length)
  return ans
};

console.log(permuteUnique([1,2,2]))